parquet-variant/src/main/java/org/apache/parquet/variant/VariantUtil.java

```diff
@@ -0,0 +1,843 @@
+    int idSize = ((typeInfo >> 2) & 0x3) + 1;
+    // Extracts b1b0 to determine the integer size of the offset list.
+    int offsetSize = (typeInfo & 0x3) + 1;
+    int idStart = pos + 1 + sizeBytes;
+    int offsetStart = idStart + numElements * idSize;
+    int dataStart = offsetStart + (numElements + 1) * offsetSize;
+    return new ObjectInfo(numElements, idSize, offsetSize, idStart, offsetStart, dataStart);
+  }
+
+  /**
+   * A helper class representing the details of a Variant array, used for `ArrayHandler`.
+   */
+  public static class ArrayInfo {
+    /** Number of object fields. */
+    public final int numElements;
+    /** The integer size of the offset list. */
+    public final int offsetSize;
+    /** The starting index of the offset list in the variant value array. */
+    public final int offsetStart;
+    /** The starting index of field data in the variant value array. */
+    public final int dataStart;
+
+    public ArrayInfo(int numElements, int offsetSize, int offsetStart, int dataStart) {
+      this.numElements = numElements;
+      this.offsetSize = offsetSize;
+      this.offsetStart = offsetStart;
+      this.dataStart = dataStart;
+    }
+  }
+
+  /**
+   * An interface for the Variant array handler.
+   * @param <T> The return type of the handler
+   */
+  public interface ArrayHandler<T> {
+    /**
+     * @param arrayInfo The details of the Variant array
+     */
+    T apply(ArrayInfo arrayInfo);
+  }
+
+  /**
+   * An interface for the Variant array handler.
+   * @param <T> The return type of the handler
+   */
+  public interface ArrayHandlerException<T> {
+    /**
+     * @param arrayInfo The details of the Variant array
+     */
+    T apply(ArrayInfo arrayInfo) throws IOException;
+  }
+
+  /**
+   * A helper function to access a Variant array, at `value[pos...]`.
+   * @param value The Variant value
+   * @param pos The starting index of the Variant value
+   * @param handler The handler to process the array
+   * @return The result of the handler
+   * @param <T> The return type of the handler
+   */
+  public static <T> T handleArray(byte[] value, int pos, ArrayHandler<T> handler) {
+    ArrayInfo info = parseArray(value, pos);
+    return handler.apply(info);
+  }
+
+  /**
+   * Same as `handleArray` but handler can throw IOException.
+   */
+  public static <T> T handleArrayException(byte[] value, int pos, ArrayHandlerException<T> handler)
+      throws IOException {
+    ArrayInfo info = parseArray(value, pos);
+    return handler.apply(info);
+  }
+
+  /**
+   * Parses the array at `value[pos...]`, and returns the array details.
+   */
+  private static ArrayInfo parseArray(byte[] value, int pos) {
+    checkIndex(pos, value.length);
+    int basicType = value[pos] & BASIC_TYPE_MASK;
+    int typeInfo = (value[pos] >> BASIC_TYPE_BITS) & PRIMITIVE_TYPE_MASK;
+    if (basicType != ARRAY) {
+      throw unexpectedType(Type.ARRAY);
```

Comment:
This will throw `MalformedVariantException`, but it is the fault of the caller that called `handleArray` and not the data. I think that this should be an `IllegalArgumentException`. The error message is fine (`Expected type to be __`).